nixos/modules/installer/tools/nixos-option/nixos-option.cc

```diff
@@ -0,0 +1,599 @@
+    std::cerr << "This attribute set contains:\n";
+    for (const auto & a : v->attrs->lexicographicOrder()) {
+        std::string name = a->name;
+        if (!name.empty() && name[0] != '_') {
+            out << name << "\n";
+        }
+    }
+}
+
+bool optionTypeIs(Context * ctx, Value & v, std::string const & sought_type)
+{
+    try {
+        auto const & type_lookup = v.attrs->find(ctx->state->sType);
+        if (type_lookup == v.attrs->end()) {
+            return false;
+        }
+        Value type = evaluateValue(ctx, type_lookup->value);
+        if (type.type != tAttrs) {
+            return false;
+        }
+        auto const & name_lookup = type.attrs->find(ctx->state->sName);
+        if (name_lookup == type.attrs->end()) {
+            return false;
+        }
+        Value name = evaluateValue(ctx, name_lookup->value);
+        if (name.type != tString) {
+            return false;
+        }
+        return name.string.s == sought_type;
+    } catch (Error &) {
+        return false;
+    }
+}
+
+Value getSubOptions(Context * ctx, Value & option)
+{
+    Value getSubOptions =
+        evaluateValue(ctx, findAlongAttrPath(*ctx->state, "type.getSubOptions", *ctx->autoArgs, option));
+    if (getSubOptions.type != tLambda) {
+        throw Error("Option's type.getSubOptions isn't a function");
+    }
+    Value emptyString{};
+    nix::mkString(emptyString, "");
+    Value v;
+    ctx->state->callFunction(getSubOptions, emptyString, v, nix::Pos{});
+    return v;
+}
+
+// Carefully walk an option path, looking for sub-options when a path walks past
+// an option value.
+Value findAlongOptionPath(Context * ctx, std::string const & path)
+{
+    Strings tokens = parseAttrPath(path);
+    Value v = ctx->options_root;
+    for (auto i = tokens.begin(); i != tokens.end(); i++) {
+        bool last_attribute = std::next(i) == tokens.end();
+        auto const & attr = *i;
+        v = evaluateValue(ctx, &v);
+        if (attr.empty()) {
+            throw Error("empty attribute name in selection path '%s'", path);
+        }
+        if (isOption(ctx, v) && optionTypeIs(ctx, v, "submodule")) {
+            v = getSubOptions(ctx, v);
+        }
+        if (isOption(ctx, v) && optionTypeIs(ctx, v, "loaOf") && !last_attribute) {
+            v = getSubOptions(ctx, v);
+            // Note that we've consumed attr, but didn't actually use it.  This is the path component that's looked up
+            // in the list or attribute set that doesn't name an option -- the "root" in "users.users.root.name".
+        } else if (v.type != tAttrs) {
+            throw Error("attribute '%s' in path '%s' attempts to index a value that should be a set but is %s", attr,
+                        path, showType(v));
+        } else {
+            auto const & next = v.attrs->find(ctx->state->symbols.create(attr));
+            if (next == v.attrs->end()) {
+                throw Error("attribute '%s' in path '%s' not found", attr, path);
```

Comment:
Not sure exactly why, but this error gets thrown with paths that contain `.`, e.g.:
```
$ nix eval '(with import <nixpkgs/nixos> {}; config.systemd.services."tinc.lugnet".serviceConfig.ExecStart)'
"/nix/store/v30a9csfl2w6wrj6lg24r09p946pvqy6-tinc-1.1pre17/bin/tincd -D -U tinc.lugnet -n lugnet -R --pidfile /run/tinc.lugnet.pid -d 0"
$ ./result/bin/nixos-option 'systemd.services."tinc.lugnet".serviceConfig.ExecStart'
error: attribute 'tinc.lugnet' in path 'systemd.services."tinc.lugnet".serviceConfig.ExecStart' not found
An error occurred while looking for attribute names. Are you sure that 'systemd.services."tinc.lugnet".serviceConfig.ExecStart' exists?
```

That said, I love that this _will_ presumably support attr names with dots in them!

Reply:
I don't have a `tinc.lugnet` service on my machine, but I do have a `environment.etc."host.conf".text`, which is a path with a dot in it.  Note that the interaction with shell quoting is somewhat confusing -- one must quote the paths to keep the shell from eating them:

```
$ ./result/bin/nixos-option environment.etc."host.conf".text
error: At 'conf' in path 'environment.etc.host.conf.text': Attribute not found

$ ./result/bin/nixos-option environment.etc.\"host.conf\".text
Value:
"multi on"

$ ./result/bin/nixos-option 'environment.etc."host.conf".text'
Value:
"multi on"
```

Reply:
Seems to be fixed in the current version, thanks!